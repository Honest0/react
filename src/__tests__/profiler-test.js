// @flow

describe('Profiler', () => {
  let React;
  let ReactDOM;
  let TestUtils;
  let store;

  const act = (callback: Function) => {
    TestUtils.act(() => {
      callback();
    });
    jest.runAllTimers(); // Flush Bridge operations
  };

  beforeEach(() => {
    store = global.store;

    React = require('react');
    ReactDOM = require('react-dom');
    TestUtils = require('react-dom/test-utils');
  });

  it('should start and stop profiling, handle root unmounting', async () => {
    const Parent = ({ count }) =>
      new Array(count).fill(true).map((_, index) => <Child key={index} />);
    const Child = () => <div>Hi!</div>;

    const containerA = document.createElement('div');
    const containerB = document.createElement('div');

    act(() => {
      ReactDOM.render(<Parent key="A" count={3} />, containerA);
      ReactDOM.render(<Parent key="B" count={2} />, containerB);
    });
    expect(store).toMatchSnapshot('1: mount');

    act(() => {
      store.startProfiling();
    });
    expect(store).toMatchSnapshot('2: profiling started');

    act(() => {
      ReactDOM.render(<Parent key="A" count={4} />, containerA);
      ReactDOM.render(<Parent key="B" count={1} />, containerB);
    });
    expect(store).toMatchSnapshot('3: update');

    act(() => ReactDOM.unmountComponentAtNode(containerB));
    expect(store).toMatchSnapshot('4: unmount B');

    act(() => ReactDOM.unmountComponentAtNode(containerA));
    expect(store).toMatchSnapshot('5: unmount A');

    act(() => {
      store.stopProfiling();
    });
    expect(store).toMatchSnapshot('6: profiling stopped');
  });
});
