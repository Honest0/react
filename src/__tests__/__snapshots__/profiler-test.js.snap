// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profiler should start and stop profiling, handle root unmounting: 1: mount 1`] = `
[root]
  ▸ <Parent key="A">
[root]
  ▸ <Parent key="B">
`;

exports[`Profiler should start and stop profiling, handle root unmounting: 2: profiling started 1`] = `
[root]
  ▸ <Parent key="A">
[root]
  ▸ <Parent key="B">
`;

exports[`Profiler should start and stop profiling, handle root unmounting: 3: update 1`] = `
[root]
  ▸ <Parent key="A">
[root]
  ▸ <Parent key="B">
`;

exports[`Profiler should start and stop profiling, handle root unmounting: 4: unmount B 1`] = `
[root]
  ▸ <Parent key="A">
`;

exports[`Profiler should start and stop profiling, handle root unmounting: 5: unmount A 1`] = ``;

exports[`Profiler should start and stop profiling, handle root unmounting: 6: profiling stopped 1`] = ``;
